src/dataflow/src/source/s3.rs

```diff
@@ -669,15 +682,20 @@ struct DownloadMetricUpdate {
 #[derive(Debug)]
 enum DownloadStatus {
```

Comment:
I think it would better if this hooked into the native Result with something like `Result<usize, DownloadError>` where `DownloadError` is

```rust
enum DownloadError {
    Retry(S3Error),
    Fail(S3Error),
    SendError,
}
```

and the `Ok(usize)` is the number of bytes downloaded or just `Result<(), DownloadError>` if you get rid of the empty record handling case (see other comment)

Reply:
Yeah I keep meaning to do that in a separate PR, we use this enum enough that it clutters up the meaningful changes shifting it in anything that changes actual code.